### ISSUE TITLE
Python package caching for gitlab

### ISSUE THREAD

**Issue by ArneTR** [green-coding-solutions/eco-ci-energy-estimation#69]
Currently the package caching for python packages is only implemented for Github.

We want to also move this functionality to Gitlab. 

If you know Gitlab better than we do and want to support, we are happy to accept a PR ❤️

If you want this feature but do not feel able to contribute please +1 so we know it is important

**Comment by bvanelli** [comment]
Just so that I understand, this issue means introducing cache on the pipeline from [eco-ci-gitlab.yml](https://github.com/green-coding-solutions/eco-ci-energy-estimation/blob/main/eco-ci-gitlab.yml). Instead of cloning the repository fresh every time, it should cache the git repo?

**Comment by ArneTR** [comment]
This is issue is actually quite outdated :)

We have migrated from Python to pure bash / CLI a while ago and thus this caching is not relevant anymore and I will close it.

However the case you bring up (cloning the GitHub Repo From GitLab) is still current. We do that bc we want to have the code centralised in one place.

In case you have a better / more efficient way idea how to do that we are super open for recommendations :) Let me know!

**Comment by bvanelli** [comment]
Yeah, we thought about this recently: if you run this pipeline on Gitlab, the repo is excluded and re-cloned on every run. 

https://github.com/green-coding-solutions/eco-ci-energy-estimation/blob/eaca3e4c511be5310fa978f6c33e5e32280c72b4/eco-ci-gitlab.yml#L26-L30

Which causes around 2.09 MiB downloaded every time. You cannot cache it right now because /tmp/ is not cacheable.

I wonder if t would be worth it to make the script cache friendly for Gitlab. There are some ways:

- Using the `cache` keyword in the job, default to caching the git clone
- Download releases (instead of cloning the repo) through a CDN, which should reduce electricity impact.
- Using a custom docker image containing the repo and a Dependency proxy.

Let me know if you think this could bring a benefit to the repo. I found it quite hard to calculate the environmental impact of the caching since some of the caching mechanisms use S3, which means that the network trip is done either way, it just offloads it somewhere else. For sure, the CDN _sounds_ more efficient, but there are [caveats](https://www.thegreenwebfoundation.org/tools/grid-aware-websites/):

> Frustratingly, CDN providers and hosting companies aren’t forthcoming in sharing information about the power utilisation of code that runs on their platforms. Because of this lack of data in the public domain, we can’t just put some code online and get power usage data back through a dashboard or API. We believe this has to change.

**Comment by ArneTR** [comment]
It technically does not have reside in /tmp. It could also be put somewhere else and on runner start we just copy it over to /tmp if necessary.
How would caching work then? GitLab would cache the FS layer internally?

2. The second thing apart from the pure external download thing is that some people have mixed feelings about download from GitHub per-se.
So I wonder if there is an option like a store in GitLab which feels more like a trustworthy download location.

Do you have experience here?

3. The environmental benefit is probably negligeable here. Since anyway the amount is very small and it is really unsure even if we use GitLab Internal caching where exactly that lives then.
I guess it is in some ways more of a feel-good thing for the user and a "likely benefit" as GitLab has quite an interest to have self managed caches close by :)

**Comment by bvanelli** [comment]
Gitlab docs describe [where the caches are stored](https://docs.gitlab.com/ci/caching/#where-the-caches-are-stored). For docker executers, they are stored locally, under docker volumes. I assume this is true for most on-prem Gitlab instances. 
The docker executor cache, a folder gets stored on a docker volume that later gets restored to the next job.

If you cache this on-prem, you skip one download per job execution, as long as the job is picked up by the same runner and the cache key is adequate.

But I also think that the gains are negligible. Using [Scope3: Power usage and embodied emissions by bandwidth (Conventional Model)](https://methodology.scope3.com/data_transfer#power-usage-and-embodied-emissions-by-bandwidth-conventional-model) as source, I get as parameters:

| Source | Unit | Value |
|--------|-----|--------|
| Electricity consumption of fixed networks	| kWh per MB	| 0.0000165 |
| Embodied emissions of fixed networks	| kgCO2e per MB	| 0.00000214 |

I live in Germany, so I'll also assume 0.334 kgCO2eq/kWh. This gives me a rate of (for both embodied and electricity carbon):

```
0.0000165 * 0.334 + 0.00000214 = 0.000007651 kgCO2e per MB
```

For the 2.2MB of the repository, and a very generous limit of 10k runs a month, for the entire year:

```
0.000007651 * 2.2 * 10000 * 12 =~ 2 kg CO2eq
```

which is the equivalent of driving 10km with a petrol car. So I guess The environmental benefit is probably negligible here, at least from my point of view.

**Comment by ArneTR** [comment]
Wow, thanks for the deep dive.

I agree. The gains here are not worth the complexity.

if we ever find a solution to eliminate the external pull and make this more "trusted" by pulling from a vetted GitLab Source like a store I think it would be worth to change the approach. But for now I would leave it as is then :)
